### PR TITLE
src/ceph-disk/ceph_disk/main.py: Make 'ceph-disk list' work on FreeBSD

### PR DESCRIPTION
**Pull request by wjwithagen**
Signed-off-by: Willem Jan Withagen <wjw@digiware.nl>

### CODE REVIEW

File: src/ceph-disk/ceph_disk/main.py [hunk 1]
```diff
@@ -762,13 +762,20 @@ def list_all_partitions():
...
     Return a list of devices and partitions
     """
-    names = os.listdir('/sys/block')
-    dev_part_list = {}
-    for name in names:
-        # /dev/fd0 may hang http://tracker.ceph.com/issues/6827
-        if re.match(r'^fd\d$', name):
-            continue
-        dev_part_list[name] = list_partitions(get_dev_path(name))
+    if not FreeBSD:
```
Comment (dachary) [inline src/ceph-disk/ceph_disk/main.py]: s/FreeBSD/FREEBSD/

Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @dachary 
Good catch

File: src/ceph-disk/ceph_disk/main.py [hunk 2]
```diff
@@ -762,13 +762,20 @@ def list_all_partitions():
...
+        names = os.listdir('/sys/block')
+        dev_part_list = {}
+        for name in names:
+            # /dev/fd0 may hang http://tracker.ceph.com/issues/6827
+            if re.match(r'^fd\d$', name):
+                continue
+            dev_part_list[name] = list_partitions(get_dev_path(name))
+    else:
+        for line in open(PROCDIR+"/partitions"):
+            columns = line.split(' ')
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: the default separator of `str.split()` is whitespace, so we can just put `line.split()` here.

File: src/ceph-disk/ceph_disk/main.py [hunk 3]
```diff
@@ -762,13 +762,20 @@ def list_all_partitions():
...
+    if not FREEBSD:
+        names = os.listdir('/sys/block')
+        dev_part_list = {}
+        for name in names:
+            # /dev/fd0 may hang http://tracker.ceph.com/issues/6827
+            if re.match(r'^fd\d$', name):
+                continue
+            dev_part_list[name] = list_partitions(get_dev_path(name))
+    else:
+        for line in open(PROCDIR+"/partitions"):
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: nit, might want to release the file object explicitly.  like
```py
with open(PROCDIR+"/partitions") as partitions:
    for line in partitions:
      # ...
```

File: src/ceph-disk/ceph_disk/main.py [hunk 4]
```diff
@@ -4524,9 +4531,35 @@ def list_devices():
...
+    except subprocess.CalledProcessError as e:
+        LOG.info('zfs list -o name,mountpoint '
+           'fails.\n (Error: %s)' % e)
+        return 1
+    lines = out.splitlines()
+    for line in lines[2:]:
+        vdevline = line.split()
+        if os.path.exists(vdevline[1]+'/active'):
+            elems = vdevline[1].split('/')
+            print(vdevline[0]+ " ceph data, active, cluster ceph, "+elems[5]+
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: pep8 warns at seeing this. maybe you can just put
```py
print(vdevline[0], "ceph data, active, cluster ceph,", elems[5],
      "mounted on:", vdevline[1])
```
?

File: src/ceph-disk/ceph_disk/main.py [hunk 5]
```diff
@@ -4524,9 +4531,35 @@ def list_devices():
...
+        out, err, ret = command(
+            [   'zfs',
+                'list',
+                '-o', 'name,mountpoint'
+            ]
+           )
+    except subprocess.CalledProcessError as e:
+        LOG.info('zfs list -o name,mountpoint '
+           'fails.\n (Error: %s)' % e)
+        return 1
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: better off raising this exception in this case, `main()` does not use the return value from `func()` or `main_catch()`, and `main_catch()` just ignore the return value of `func()`.

File: src/ceph-disk/ceph_disk/main.py [hunk 6]
```diff
@@ -4524,9 +4531,35 @@ def list_devices():
...
+    lines = out.splitlines()
+    for line in lines[2:]:
+        vdevline = line.split()
+        if os.path.exists(vdevline[1]+'/active'):
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]:
```py
os.path.exists(os.path.join(vdevline[1], 'active')
```

File: src/ceph-disk/ceph_disk/main.py [hunk 7]
```diff
@@ -4524,9 +4531,35 @@ def list_devices():
...
+            elems = vdevline[1].split('/')
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]:
```py
elems = os.path.split(vdevline[1])
```

File: src/ceph-disk/ceph_disk/main.py (hunk 4 above)
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: and there are other places that violates pep8 in this change, could you run this file with `pep8`, and make sure it's pep8 warning/errors free?

**Comment by wjwithagen** [comment]
@tchaikov @dachary 
I'll look into all the remarks, and questions. Given my lack of experience with Python that migt take a few tries. So please have pantience with me.

File: src/ceph-disk/ceph_disk/main.py (hunk 5 above)
Comment (wjwithagen) [inline src/ceph-disk/ceph_disk/main.py]: @tchaikov 
Removed the return. 
But if listing fails, there is not much more we can do?

Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: we should re-raise the exception. so `main_catch()` can catch it, or, user can get more clue pass `--verbose` to ceph-disk at seeing the error, which will print out the backtrace in any exception is raised.

File: src/ceph-disk/ceph_disk/main.py [hunk 8]
```diff
@@ -4524,9 +4527,35 @@ def list_devices():
...
+        out, err, ret = command(
+            [
+                'zfs',
+                'list',
+                '-o', 'name,mountpoint'
+            ]
+        )
+    except subprocess.CalledProcessError as e:
+        LOG.info('zfs list -o name,mountpoint '
+                 'fails.\n (Error: %s)' % e)
```
Comment (tchaikov) [inline src/ceph-disk/ceph_disk/main.py]: need to re-raise this exception, you can just add
```py
raise
```
in the `except` block.

**Comment by wjwithagen** [comment]
@tchaikov @dachary 
Issues should be addressed, can you take another look and/or set QA

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-04-18_05:51:22-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2017-04-18_08:13:21-rados-wip-kefu-testing---basic-mira/

**Comment by tchaikov** [comment]
being tested at http://pulpito.ceph.com/kchai-2017-04-18_15:16:55-ceph-disk-wip-kefu-testing---basic-vps/

**Comment by tchaikov** [comment]
@dachary did @wjwithagen address your comment?

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-04-18_15:16:55-ceph-disk-wip-kefu-testing---basic-vps/

**Pull request merged**